### ISSUE TITLE
Poor performance on surviv.io

### ISSUE THREAD

**Issue by Ceremony64** [marklieberman/foxygestures#232]
I am getting very low performance on surviv.io: Whenever I click in the game with this extension enabled, the screen freezes for a few frames. Once I disable the extension, clicking no longer causes any slowdowns.
These slowdowns occur for both left- and rightclick, no matter the selected gesture button.

P.S. Tested with a clean FF profile

**Comment by marklieberman** [comment]
The problem is here: https://github.com/marklieberman/foxygestures/blob/bafdcfaad8b6c666942e97d3611fbc544123a913/src/content/helpers.js#L173

It copies the canvas data to a data:// URL on mousedown which is what introduces the lag. However, if I remove it then it breaks gestures like Save As that save the canvas image to a file.

**Comment by Ceremony64** [comment]
canvas.toDataURL() encodes the image as PNG, which is really slow, especially for large canvases like in this game.
Instead, use `canvas.getContext("2d").getImageData(0,0,canvas.width,canvas.height)` to cache the raw image data and only convert it to PNG when needed (by drawing it to an offscreen canvas and calling toDataURL()). Speedup should be somewhere around 25x to 100x.

Also, this should only be triggered by the assigned mouse button: I use gestures with the right click, but it also triggers on left click.

Skipping the canvas cache entirely when the save-canvas features is not assigned (which it isn't by default) should also help, tho I'm not sure how to take the custom userscripts into account.

**Comment by marklieberman** [comment]
I didn't know about the getImageData thing so I'll implement that. I can also short-circuit the mousedown handler for left-click if rocker-gestures are not enabled as well.

Unfortunately, I don't have much time to work on FG in the next 2 months (missed a lot of work due to cancer) so a PR will be appreciated if you are willing.

**Comment by mathew832** [comment]
Looking forward to this being remedied since I suspect it's causing my issues on Google maps when clicking to drag. When clicking to drag the map around, It takes almost a near second for anything to happen. Google maps works fine with the extension disabled though. 

But no hurry. Hope you get better soon and thanks for the great extension. Been using almost ever since v57 came out.

**Comment by marklieberman** [comment]
@Ceremony64 I was just looking at this issue, and I noticed that google maps and surviv.io use a webgl rendering context rather then 2d. The getImageData() method isn't available on webgl contexts and furthermore, when preserveDrawingBuffer = false, toDataURL() returns a transparent image. I think on that basis I will not support saving images from webgl canvas'.

I have made it such that the canvas data for 2d canvas' is fetched asynchronously as just before saving. It's a complicated workflow that cannot be implemented as one long promise, but I guess I'll take that trade for the performance boost when dealing with canvases.
